onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -1039,22 +1118,37 @@ Applies to session load, initialization, etc. Default is 0.)pbdoc")
           "Rpbdoc(Specify values of named dimensions within model inputs.)pbdoc")
       .def(
           "add_session_config_entry",
-          [](SessionOptions* options, const char* config_key, const char* config_value) -> void {
+          [](PySessionOptions* options, const char* config_key, const char* config_value) -> void {
             const Status status = AddSessionConfigEntryImpl(*options, config_key, config_value);
             if (!status.IsOK())
               throw std::runtime_error(status.ErrorMessage());
           },
           "Rpbdoc(Set a single session configuration entry as a pair of strings.)pbdoc")
       .def(
           "get_session_config_entry",
-          [](SessionOptions* options, const char* config_key) -> std::string {
+          [](PySessionOptions* options, const char* config_key) -> std::string {
             const std::string key(config_key);
             if (!HasSessionConfigEntry(*options, key))
               throw std::runtime_error("SessionOptions does not have configuration with key: " + key);
 
             return options->session_configurations.at(key);
           },
-          "Rpbdoc(Get a single session configuration value using the given configuration key.)pbdoc");
+          "Rpbdoc(Get a single session configuration value using the given configuration key.)pbdoc")
+      .def(
+          "register_custom_ops_library",
+          [](PySessionOptions* options, const std::string& library_path)
+              -> void {
+            OrtSessionOptions s;
```

Comment:
Do we've to do all the following processing at this stage? The session options is a lightweight container of all the options. We can save just a vector of library paths in the session options and then when the inference session is constructed, we can load the libraries and extract the domains. This will avoid having to store shared_ptr of custom_op_libraries in 2 different places; it can simply be stored in one place as a unique_ptr (in the PyInferenceSession) and the ownership (and unloading) semantics become clearer.

Reply:
Per offline sync, keeping it as it is, documenting why we do it the way we do